crates/gitbutler-command-context/src/context.rs

```diff
@@ -0,0 +1,174 @@
+        // XXX(qix-): We do this because the internal repository we use to store the "virtual"
+        // XXX(qix-): refs and information use Git's alternative-objects mechanism to refer
+        // XXX(qix-): to the project repository's objects. However, the project repository
+        // XXX(qix-): has no knowledge of these refs, and will GC them away (usually after
+        // XXX(qix-): about 2 weeks) which will corrupt the internal repository.
+        // XXX(qix-):
+        // XXX(qix-): We will ultimately move away from an internal repository for a variety
+        // XXX(qix-): of reasons, but for now, this is a simple, short-term solution that we
+        // XXX(qix-): can clean up later on. We're aware this isn't ideal.
+        if let Ok(config) = repo.config().as_mut() {
+            let should_set = match config.get_bool("gitbutler.didSetPrune") {
+                Ok(false) => true,
+                Ok(true) => false,
+                Err(err) => {
+                    tracing::warn!(
+                                "failed to get gitbutler.didSetPrune for repository at {}; cannot disable gc: {}",
+                                project.path.display(),
+                                err
+                            );
+                    false
+                }
+            };
+
+            if should_set {
+                if let Err(error) = config
+                    .set_str("gc.pruneExpire", "never")
+                    .and_then(|()| config.set_bool("gitbutler.didSetPrune", true))
+                {
+                    tracing::warn!(
+                                "failed to set gc.auto to false for repository at {}; cannot disable gc: {}",
+                                project.path.display(),
+                                error
+                            );
+                }
+            }
+        } else {
+            tracing::warn!(
+                "failed to get config for repository at {}; cannot disable gc",
+                project.path.display()
+            );
+        }
+
+        if repo.on_integration_branch()? {
+            Ok(RequestContext::OpenWorkspace(OpenWorkspaceContext {
+                git_repository: repo,
+                project: project.clone(),
+            }))
+        } else {
+            Ok(RequestContext::OutsideWorkspace(OutsideWorkspaceContext {
+                git_repository: repo,
+                project: project.clone(),
+            }))
+        }
+    }
+
+    pub fn try_create_open_workspace_context(project: &Project) -> Result<OpenWorkspaceContext> {
+        let RequestContext::OpenWorkspace(open_workspace_context) = RequestContext::open(project)?
+        else {
+            bail!("Open workspace required for this action");
+        };
+
+        Ok(open_workspace_context)
+    }
+}
+
+pub trait ContextProjectAccess {
+    fn project(&self) -> &Project;
+    fn set_project(&mut self, project: &Project);
+}
+
+pub trait ContextRepositoryAccess {
```

Comment:
I think anything that works with this trait now could instead just work with a `&git2::Repository` - that's a great finding that would simplify the functions. Once they need more, that can be changed.

This also means the caller invokes the `.repo()` method to produce the required `git2::Repository`.